src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    private Set<String> initialLabels;
+    private List<String> addedLabels;
+    private List<String> removedLabels;
+    private List<String> matchedLabels;
+    private List<TurboLabel> allLabels;
+    private OptionalInt currentSuggestionIndex;
+
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> allLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), allLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> allLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.allLabels = allLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private final void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            updateIfMatchesLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Updates current state if there is at least one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private final void updateIfMatchesLabel(String keyword) {
+        if (TurboLabel.hasMatchedLabel(allLabels, keyword)) {
+            updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(allLabels, keyword));
+        }
```

Comment:
I agree that we should avoid complicating methods. However, lower level method should be solid because they are being called from many other places. `updateAssignedLabels` is not solid because it assumes the parameter is not null but doesn't verify that. We can add an `assert` to verify that, but that still means every call site is responsible for the null check. If we make the method respond to null (in this case, null means 'no label') in the right way, call sites don't have to do the check many times.
No, we don't have to name it `updateAssignedLabelsIfParameterIsPresent`because if the parameter is declared Optional, 'nothing will be updated if the parameter is not present' can be implied.